src/afl-fuzz-statsd.c

```diff
@@ -0,0 +1,158 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <string.h>
+#include <sys/types.h>
+#include <netdb.h>
+#include <unistd.h>
+#include "afl-fuzz.h"
+
+#define MAX_STATSD_PACKET_SIZE 4096
+#define MAX_TAG_LEN 200
+#define METRIC_PREFIX "fuzzing"
+
+int statsd_socket_init(afl_state_t *afl) {
+
+  /* Default port and host.
+  Will be overwritten by AFL_STATSD_PORT and AFL_STATSD_HOST environment
+  variable, if they exists.
+  */
+  u16   port = STATSD_DEFAULT_PORT;
+  char *host = STATSD_DEFAULT_HOST;
+
+  if (afl->afl_env.afl_statsd_port) {
+
+    port = atoi(afl->afl_env.afl_statsd_port);
+
+  }
+
+  if (afl->afl_env.afl_statsd_host) { host = afl->afl_env.afl_statsd_host; }
+
+  int sock;
+  if ((sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) == -1) {
+
+    FATAL("Failed to create socket");
+
+  }
+
+  memset(&afl->statsd_server, 0, sizeof(afl->statsd_server));
+  afl->statsd_server.sin_family = AF_INET;
+  afl->statsd_server.sin_port = htons(port);
+
+  struct addrinfo *result;
+  struct addrinfo  hints;
+
+  memset(&hints, 0, sizeof(struct addrinfo));
+  hints.ai_family = AF_INET;
+  hints.ai_socktype = SOCK_DGRAM;
+
+  if ((getaddrinfo(host, NULL, &hints, &result))) {
+
+    FATAL("Fail to getaddrinfo");
+
+  }
+
+  memcpy(&(afl->statsd_server.sin_addr),
+         &((struct sockaddr_in *)result->ai_addr)->sin_addr,
+         sizeof(struct in_addr));
+  freeaddrinfo(result);
+
+  return sock;
+
+}
+
+int statsd_send_metric(afl_state_t *afl) {
+
+  char buff[MAX_STATSD_PACKET_SIZE] = {0};
+
+  /* afl->statsd_sock is set once in the initialisation of afl-fuzz and reused
+  each time If the sendto later fail, we reset it to 0 to be able to recreates
+  it.
+  */
+  if (!afl->statsd_sock) {
+
+    afl->statsd_sock = statsd_socket_init(afl);
+    if (!afl->statsd_sock) {
+
+      WARNF("Cannot create socket");
+      return -1;
+
+    }
+
+  }
+
+  statsd_format_metric(afl, buff, MAX_STATSD_PACKET_SIZE);
+  if (sendto(afl->statsd_sock, buff, strlen(buff), 0,
+             (struct sockaddr *)&afl->statsd_server,
+             sizeof(afl->statsd_server)) == -1) {
+
+    if (!close(afl->statsd_sock)) { perror("Cannot close socket"); }
+    afl->statsd_sock = 0;
+    WARNF("Cannot sendto");
+    return -1;
+
+  }
+
+  return 0;
+
+}
+
+int statsd_format_metric(afl_state_t *afl, char *buff, size_t bufflen) {
+
+/* Metric format:
+<some.namespaced.name>:<value>|<type>
+*/
+#ifdef USE_DOGSTATSD_TAGS
```

Comment:
Are these tags only used by Datadog or supported by other StatsD providers?
Not a big fan of a for-profit kinda-proprietary format here, maybe we find a way to make this generalizable/configurable?

Reply:
Ah the linked comment, https://github.com/netdata/netdata/issues/3813#issuecomment-399276423, kinda answers my question. Maybe we don't need the ifdef at all, other consumers could just ignore the tag?

Reply:
I agree, and it is possible to add other format as well.

(from the PR comment directly, the links are working there, not in this quote)
> The code supports metric tags (fixed to banner and afl_version for now), with the |#tag:value notation.
Since there is a no standards for tags, I only implemented DogstatsD format, since that it seems to be the "first" one based on this (I didn't verified) and the one that I will be using / testing. If desired, I can add support for other through a config (#define USE_STATSD_INFLUXDB for example) in another PR.
For more information about different format, see prometheus/statsd_exporter or statsd issue #619

I just wanted to push the PR first, before implementing other format if the PR wasn't "accepted".

From what I saw, there is total  incompatibility between each provider, so will need to add a few common ones. Not sure which should be the default one though.

You can see a list [there](https://github.com/prometheus/statsd_exporter#tagging-extensions)

Reply:
Okay, the way I see it it's mostly the seperator that differs, the content stays the same?
So we can easily set an sprintf string in the config (or via env?) the content stays the same, right?

Reply:
Okay, the way I see it it's mostly the seperator that differs, the content stays the same?
So we can easily set an sprintf string in the config (or via env?) the content stays the same, right?
/edit:
Or do we just want to add the 4 or 5 supported ones and have a `AFL_STATSD_TAG_FLAVOR` env
(there is also the grahite flavor missing from the list: https://graphite.readthedocs.io/en/latest/tags.html )

Reply:
Yep, that's what I'm thinking too.

So, changing the #define by an env var and then doing something along the lines of 
```c
else if (AFL_STATSD_TAG_FLAVOR == USE_LIBRATO_TAGS) {
 char tags[MAX_TAG_LEN * 2] = {0};
  snprintf(tags, MAX_TAG_LEN * 2, "#banner=%s,afl_version=%s", afl->use_banner,
           VERSION);
} else if  (AFL_STATSD_TAG_FLAVOR == USE_INFLUXDB_TAGS) {
 char tags[MAX_TAG_LEN * 2] = {0};
  snprintf(tags, MAX_TAG_LEN * 2, ",banner=%s,afl_version=%s", afl->use_banner,
           VERSION);
} else if  (AFL_STATSD_TAG_FLAVOR == USE_SIGNALFX_TAGS) {
 char tags[MAX_TAG_LEN * 2] = {0};
  snprintf(tags, MAX_TAG_LEN * 2, "[banner=%s,afl_version=%s]", afl->use_banner,
           VERSION);
}
```

The annoyance is that they aren't injected in the same position in the metrics.
DogstatsD is at the end of the line while most of the other are just after the metric name.
But that's not a big deal, for sure.

Reply:
ooh different locatons? Well that's just mean.
I was thinking along the lines of (during init)
```c
if (AFL_STATSD_TAG_FLAVOR == LIBRATO) {
  afl_state->statsd_format = "#banner=%s,afl_version=%s"
} else ...
```
And then later using `afl_state->statsd_format`in `snprintf` directly...

That being said, if the location of the tag differs anyway, you could also have individual complete snprintfs (the whole packet), hidden behind the if of your example above - and play around with some preprocessor magic to build the string and params? could be a bit define heavy  but quick - up to you :)

Reply:
> From what I saw, there is total incompatibility between each provider, so will need to add a few common ones. Not sure which should be the default one though.
the default should be a non-commercial one.
Also it is fine if you add support for other variants if you want. we will accept the PR :)

Reply:
Ok, so I made a few modification but it should be good now. I set the default to no tag if there is no/empty`AFL_STATSD_TAGS_FLAVOR` provided.

Quite unsure if the init should be done in `afl-fuzz.c:895`, but that looked a good place for it.
I inspected all the packets from the different types of tags type (with Wireshark, I don't have any other platform set up):
(truncated to first 4 lines each)

`AFL_STATSD_TAGS_FLAVOR=`:
```
fuzzing.cycle_done:0|g
fuzzing.cycles_wo_finds:0|g
fuzzing.execs_done:95077|g
fuzzing.execs_per_sec:4128.22|g
...
```

`AFL_STATSD_TAGS_FLAVOR=librato`:
```
fuzzing.cycle_done#banner=f1,afl_version=++3.00a:0|g
fuzzing.cycles_wo_finds#banner=f1,afl_version=++3.00a:0|g
fuzzing.execs_done#banner=f1,afl_version=++3.00a:9|g
fuzzing.execs_per_sec#banner=f1,afl_version=++3.00a:9000.00|g
...
```

`AFL_STATSD_TAGS_FLAVOR=dogstatsd`:
```
fuzzing.cycle_done:0|g|#banner:f1,afl_version:++3.00a
fuzzing.cycles_wo_finds:0|g|#banner:f1,afl_version:++3.00a
fuzzing.execs_done:77557|g|#banner:f1,afl_version:++3.00a
fuzzing.execs_per_sec:4136.37|g|#banner:f1,afl_version:++3.00a
...
```

`AFL_STATSD_TAGS_FLAVOR=signalfx`:
```
fuzzing.cycle_done[banner=f1,afl_version=++3.00a]:0|g
fuzzing.cycles_wo_finds[banner=f1,afl_version=++3.00a]:0|g
fuzzing.execs_done[banner=f1,afl_version=++3.00a]:4177|g
fuzzing.execs_per_sec[banner=f1,afl_version=++3.00a]:4135.64|g
...
```

`AFL_STATSD_TAGS_FLAVOR=influxdb`:
```
fuzzing.cycle_done,banner=f1,afl_version=++3.00a:0|g
fuzzing.cycles_wo_finds,banner=f1,afl_version=++3.00a:0|g
fuzzing.execs_done,banner=f1,afl_version=++3.00a:4164|g
fuzzing.execs_per_sec,banner=f1,afl_version=++3.00a:4147.41|g
...
```